### DISCUSSION TITLE
Allegro LAMMPS on AMD GPUs

### DISCUSSION THREAD

**Discussion by johkl** [mir-group/pair_nequip_allegro#23] [General]
Dear developers,

Thanks for the excellent work and set of codes!

Do you have, or know of anyone else who have, any success in using/porting the pair_allegro LAMMPS patch on AMD GPUs? and if so what have been their experience?

Best,
Johan

**Comment by Linux-cpp-lisp** [comment]
Hi Johan,

Thanks for your your interest in our code!

Yes, we know of at least one or two groups that have done this with some success.  My general understanding is that it didn't require too much fiddling, since ROCm devices are exposed at a user level as CUDA devices in PyTorch. I'll ping @svandenhaute in case he's willing to share his experience.

**Reply by svandenhaute** [reply]
In general, it's been quite smooth! I was working on LUMI, with a system-provided MPICH and AMD MI250X (8 effective GPUs per node). Apart from telling Kokkos that you'll be using HIP/ROCm and in addition the GPU version (e.g sm80 for Nvidia A100), there was nothing really particular about it!

**Reply by johkl** [reply]
Thank you both for the quick replies! Good to know that it should be fairly straightforward to set up!

**Reply by Linux-cpp-lisp** [reply]
Thanks @svandenhaute !

**Reply by tgmaxson** [reply]
@svandenhaute Out of curiosity, how fast was it relative to an A100?  I successfully got a 6800XT running, but the performance was terrible but this was likely due to the driver configuration being unofficially supported (but served its purpose for comparing hardware types).

**Reply by svandenhaute** [reply]
If I remember correctly, the performance difference between a single A100 and a single MI250x GCD was relatively small, if any. I'm not sure whether we'd expect one to be significantly faster than the other based on memory bandwidth and/or cuda core count, but with ROCm 5.2.3 / CUDA 11.3, they were certainly comparable.
Switching to a more recent ROCm version might significantly speed up the AMD performance (?)

**Reply by johkl** [reply]
If I may ask @svandenhaute, what pytorch/libtorch version did you use?

**Reply by svandenhaute** [reply]
I used libtorch-cxx11-abi-shared-with-deps-1.13.0+rocm5.2. If the ROCm on your system is newer, this version will most likely not work because in contrast to CUDA, ROCm 5.2 --> 5.3 introduced breaking changes in terms of where it's supposed to look for certain shared libraries -- based on my experience at least.

**Reply by johkl** [reply]
Thanks @svandenhaute !

Using that same libtorch version and a system ROCm 5.0.2. I'm getting a compilation error (shorting the file paths for simplicity): 

In file included from ..../lammps/src/KOKKOS/pair_allegro_kokkos.cpp:36:
In file included from ..../libtorch/include/c10/cuda/CUDACachingAllocator.h:4:
In file included from ..../libtorch/include/c10/cuda/CUDAGraphsC10Utils.h:3:
...../libtorch/include/c10/cuda/CUDAStream.h:6:10: fatal error: 'cuda_runtime_api.h' file not found

Cmake cmd is the following:

cmake ../cmake -DCMAKE_PREFIX_PATH=..../libtorch_rocm5.2_1.13/libtorch -DPKG_KOKKOS=ON -DCMAKE_CXX_COMPILER=hipcc  -D Kokkos_ARCH_VEGA90A=ON -D Kokkos_ENABLE_HIP=ON -D Kokkos_ENABLE_CUDA=OFF -D Kokkos_ENABLE_OPENMP=ON  -DMKL_INCLUDE_DIR="$CONDA_PREFIX/include"

Any ideas for a solution would be appreciated!

**Reply by svandenhaute** [reply]
No idea to be honest. @anjohan ?

**Reply by johkl** [reply]
Turns out I was on the "multicut" branch. That error disappears on the "stress" branch.

Just in the unlikely case that someone is has seen this before; I am getting a different error now: 

No rule to make target '/opt/rocm/hip/lib/libamdhip64.so', needed by 'lmp'.

Which is due to the fact that the correct target path should "/opt/rocm-5.0.2/... " rather than "/opt/rocm/..." but I can't seem to find where/how CMake gives the instruction to look in "/opt/rocm/..." ...

Sorry for the spam!

**Reply by svandenhaute** [reply]
Are you sure `libamdhip64.so` is actually provided by ROCm 5.0.2? I seem to remember that they renamed that library or moved it into some other file. If you have the ability to try a newer ROCm, I think that's your best bet.

**Reply by johkl** [reply]
The libamdhip64.so did indeed exist and I did mange to make it compile now using ROCm 5.0.2.  I had to manually change all "/opt/rocm/..." paths to "/opt/rocm-5.0.2/..." in build/CMakeFiles/lmp.dir/build.make and build/CMakeFiles/lmp.dir/link.txt . I'm a cmake novice, so I don't know why it choose to look for certain libraries in /opt/rocm/ and others in /opt/rocm-5.0.2/ .

**Reply by Linux-cpp-lisp** [reply]
I've never done this with ROCm, but I've had to fix similar problems in the recent past with MKL by doing things like `-DCMAKE_LIBRARY_PATH=$MKL_LIB -DCMAKE_INCLUDE_PATH=$MKL_INCLUDE` where those environment variables are provided by my system.  Note I didn't copy paste those options, that's from memory, so more an idea than something to directly try in your command.

**Reply by Linux-cpp-lisp** [reply]
Here's what it was for MKL:
```
-DCMAKE_PREFIX_PATH="$MKLROOT;$CONDA_PREFIX" -DCMAKE_LIBRARY_PATH=$MKL_LIB
```

**Reply by johkl** [reply]
I did (very non-systematically) try adding appropriate paths using combinations of CMAKE_LIBRARY_PATH / CMAKE_INCLUDE_PATH / CMAKE_PREFIX_PATH . 

I suppose it should definitely be possible to solve the problem in this way, I just never managed to fully do it this way.

**Reply by hmcezar** [reply]
Sorry to come back to this after so long, but I'm trying to create a container with NequIP/Allegro/LAMMPS on LUMI and I thought maybe @svandenhaute or @Linux-cpp-lisp could help.

Using `LUMI/23.03` I managed to create a container with `conda-containerize` that allowed me to install NequIP, Allegro and Torch 1.11.0 (from source).
But I'm having problems with `pair_allegro` now.

After getting LAMMPS `stable_2Aug2023_update2` and applying the patch from the `multicut` branch I had to manually add 
```cmake
find_package(hip REQUIRED)
find_package(rocblas REQUIRED)
find_package(hipfft REQUIRED)
find_package(hipsparse REQUIRED)
find_package(hiprand REQUIRED)
```
to `lammps/cmake/CMakeLists.txt` to pass `cmake`.
However, when I run `make` I'm getting:
```makefile
/project/project_465000259/musselic/software/pytorch/torch/include/c10/macros/Macros.h:368:5: error: __host__ __device__ function '__assert_fail' cannot overload __device__ function '__assert_fail'
    __assert_fail(
    ^
/opt/rocm-5.2.3/include/hip/amd_detail/amd_device_functions.h:832:6: note: previous declaration is here
void __assert_fail(const char *assertion,
     ^
```

Any ideas on how to solve this? I feel that I'm close, but I'm not sure how to move forward.

**Reply by Linux-cpp-lisp** [reply]
@anjohan ?

**Reply by hmcezar** [reply]
So, after digging a bit (and checking a few git blames) I think the issue is I'm using pytorch 1.11.0, which had this issue:
https://github.com/pytorch/pytorch/pull/73040

In this case I think I'll have to patch pytorch or use a version that is not 1.11.0.

Is 1.11.0 still the recommended version for production runs @Linux-cpp-lisp ?

If any of you have any suggestions, I'm open to new ideas.

Thanks!

**Reply by Linux-cpp-lisp** [reply]
Oh, right---on AMD GPUs, specifically, I'm pretty sure later versions are fine. (See https://github.com/mir-group/nequip/discussions/311#discussioncomment-5231630 and Ctrl-F AMD)
